### ISSUE TITLE
Batching mutations?

### ISSUE THREAD

**Issue by haf** [mrange/fsharpadvent2016#2]
Similar to [Immutable.JS](https://facebook.github.io/immutable-js/docs/#/Map/withMutations), perhaps?

```js
var map1 = Immutable.Map();
var map2 = map1.withMutations(map => {
  map.set('a', 1).set('b', 2).set('c', 3);
});
assert(map1.size === 0);
assert(map2.size === 3);
```

**Comment by mrange** [comment]
I know that clojure map and Fsharpx.collections has destructive insertions but when I tested them out I didn't seem to gain performance on insert and lost some on lookup so I didn't add it yet.

The lookup loss of performance seemed due to that the tries seemed more "holey" (to leave room for new nodes) which cost something.

Insertion should be faster but I didn't see it.

So I wasn't motivated at the time. 

With that said; I will think a bit on it to see how it can be added.

**Comment by haf** [comment]
My use-case is to mutate all values in the map. Is there a better pattern for doing this?

BTW; it's being put to good use here https://github.com/logary/logary/pull/219

**Comment by mrange** [comment]
Well if you like to mutate all values that would mean you like a new trie with identical structure that has new values. That is basically a clone with value replacement

That should be very quick but how should the signature look?

```fsharp
val mapValues : ('K -> 'V -> 'U) -> PersistentHashMap<'K, 'V> -> PersistentHashMap<'K, 'U>
```

**Comment by mrange** [comment]
Would a function like above work?

**Comment by haf** [comment]
Yes, that would work, but wouldn't it be inconsistent with yielding `KeyValuePair<'K, 'V>` everywhere else? `:PHM :> _ seq` for example. Although KVP is a struct so I can see why it's a good fit.

So yes.

**Comment by mrange** [comment]
The map function is in curried form which I think makes it ok for consistency. As for `KeyValuePair<'K, 'V>` I picked that over `'K*'V` because `KeyValuePair<'K, 'V>` is a struct (reducing GC pressure).

**Comment by mrange** [comment]
The implementation of above function should be "trivial" and perform well.

**Comment by mrange** [comment]
@haf, master now has a proposal for you.

**Comment by haf** [comment]
Thank you Mårten, I'll try them out =)
